spec/Candid.md

```diff
@@ -882,200 +882,150 @@ service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>;
-
--------------------
-true ~> true : bool
-
----------------------
-false ~> false : bool
-
------------------------
-<text> ~> <text> : text
-
--------------------
-null ~> null : null 
+C[<t> <: <t>](x) = x    for every <t> ∈ <numtype>, bool, text, null
 ```
 
-Values of type `nat` coerce successfully at type `int`:
+Values of type `nat` coerce at type `int`:
 ```
---------------------------------
-(<x> : nat) ~> (<x> : int) : int
+C[nat <: int](<nat>) = <nat>
 ```
 
-Any value coerces at type `reserved`, producing the canonical value of type `reserved`:
-
+Coercion into `reserved` is the constant map (this is arbitrarily using `null` as “the” value of `reserved`):
 ```
------------------------------------
-<v> ~> (null : reserved) : reserved
+C[<t> <: reserved](_) = null
 ```
-
-NB: No rule is needed for type `empty`, because there are no values of that type. By construction, `<v> ~> _ : empty` will not hold.
+NB: No rule is needed for type `empty`, because there are no values of that type. By construction, `C[empty <: <t>]` is the unique function on the empty domain.
 
 #### Vectors
 
-Only vectors coerce at vector types, and only if all elements coerce successfully.
-
+Vectors coerce pointwise:
 ```
-<v> ~> <v'> : <t>
------------------------------------------
-vec { <v>;* } ~> vec { <v'>;* } : vec <t>
+C[vec <t> <: vec <t'>]( vec { <v>;* } ) = vec { C[<t> <: <t'>](<v>);* }
 ```
 
 #### Options
 
-The null value coerce at any option type:
+The null type coerces into any option type:
 ```
-----------------------
-null ~> null : opt <t>
+C[null <: opt <t>](null) = null
 ```
 
-An optional value coerces at an option type, if the constituent value coerces at the constituent type:
-```
-<v> ~> <v'> : <t>
------------------------------
-opt <v> ~> opt <v'> : opt <t>
+An optional value coerces at an option type, if the constituent value has a suitable type, and else goes to `null`:
 ```
-
-If an optional value _fails_ to coerce at an optional type, or the value is `reserved`, the result is `null`, not failure:
+C[opt <t> <: opt <t'>](null) = null
+C[opt <t> <: opt <t'>](opt <v>) = opt C[<t> <: <t'>](v)  if <t> <: <t'>
+C[opt <t> <: opt <t'>](opt <v>) = null                   if not(<t> <: <t'>)
 ```
-not (<v> ~> _ : <t>)
--------------------------
-opt <v> ~> null : opt <t>
 
------------------------------------
-(null : reserved) ~> null : opt <t>
-```
+NOTE: These rules above imply that a Candid decoder has to be able to decide the subtyping relation at runtime.
 
 Coercing a non-null, non-optional and non-reserved value at an option type treats it as an optional value:
 ```
-<v> ≠ null
-<v> ≠ (null : reserved)
-<v> ≠ opt _
-not (null <: <t>)
-opt <v> ~> <v'> : opt <t>
--------------------------
-<v> ~> <v'> : opt <t>
+C[<t> <: opt <t'>](<v>) = opt C[<t> <: <t'>](v)  if not (null <: <t>) and <t> <: <t'>
+C[<t> <: opt <t'>](_)   = null                   if not (null <: <t>) and not (<t> <: <t'>)
```

Comment:
```suggestion
C[<t> <: opt <t'>](_)   = null                   otherwise
```
We don't have a rule for `null <: t'` then. I think this applies to the `null <: t'` case as well?

Reply:
I think you are right. But let’s be explicit here, and and not rely on fall-through-semantics.

So it seems the fomulation that we want, so that each line is correct independent of their order, might be

```
C[<t> <: opt <t'>](<v>) = opt C[<t> <: <t'>](v)  if not (null <: <t'>) and <t> <: <t'>
C[<t> <: opt <t'>](_)   = null                   if not (null <: <t'>) and not (<t> <: <t'>)
C[<t> <: opt <t'>](_)   = null                   if null <: <t'> and not (null <: <t>)
```

(Remember that `null <: t` implies that `t = null`, `t = opt …` or `t = reserved`, and these cases have their own rules)


I think this needs a table:

| null <: t | null <: t' | t <: t' | Rule?
| --- | --- | --- | --- |
| no | no | no | second rule
| no | no | yes | first rule
| no | yes | no | third rule
| no | yes | yes | third rule
| yes | no | no | second rule
| yes | no | yes | impossible (<: is transitive)
| yes | yes | no | see rules for  `t = null`, `t = opt …` or `t = reserved`
| yes | yes | yes | see rules for  `t = null`, `t = opt …` or `t = reserved`